src/app/icd/server/ICDMonitoringTable.cpp

```diff
@@ -263,16 +268,26 @@ CHIP_ERROR ICDMonitoringTable::Set(uint16_t index, const ICDMonitoringEntry & en
     VerifyOrReturnError(entry.keyHandleValid, CHIP_ERROR_INVALID_ARGUMENT);
 
     ICDMonitoringEntry e(this->mFabric, index);
-    e.checkInNodeID    = entry.checkInNodeID;
-    e.monitoredSubject = entry.monitoredSubject;
-    e.clientType       = entry.clientType;
-    e.index            = index;
+    e.checkInNodeID     = entry.checkInNodeID;
+    e.monitoredSubject  = entry.monitoredSubject;
+    e.clientType        = entry.clientType;
+    e.index             = index;
+    e.symmetricKeystore = entry.symmetricKeystore;
 
     memcpy(e.aesKeyHandle.AsMutable<Crypto::Symmetric128BitsKeyByteArray>(),
            entry.aesKeyHandle.As<Crypto::Symmetric128BitsKeyByteArray>(), sizeof(Crypto::Symmetric128BitsKeyByteArray));
     memcpy(e.hmacKeyHandle.AsMutable<Crypto::Symmetric128BitsKeyByteArray>(),
            entry.hmacKeyHandle.As<Crypto::Symmetric128BitsKeyByteArray>(), sizeof(Crypto::Symmetric128BitsKeyByteArray));
 
+    ReturnErrorOnFailure(e.symmetricKeystore->PersistICDKey(e.aesKeyHandle));
```

Comment:
> in `ICDMonitoringTable::Remove`, `entry.DeleteKey()` is called and the key is destroyed. I've checked if there are no potential key leaks and couldn't spot anything. During some smoke tests, keys have been destroyed correctly as well. Can you see any more cases when the key should be destroyed but it itsn't now?

Ah, right. After double-checking it looks correct.

> I thought about that, but for ICD entries, `index` is always shifted when entry is removed (see https://github.com/project-chip/connectedhomeip/blob/master/src/app/icd/server/ICDMonitoringTable.cpp#L288). So If I wanted to assign the key with `index`, we would need to remove that mechanism in `Remove` and add a similar one to current search for free key ID in `Set` (we would still need to find next free index). I've decided to just implement `FindFreeKeySlotInRange` as it is quite universal and could come in handy in the future if "dynamic" PSA key slot allocation would be needed somewhere else

Hmm. Even if you derived the key ID from `index`, the shifting mechanism would work correctly because during the shift `ICDMonitoringTable::Set` would be called, which would _move_ the corresponding PSA keys into the new IDs, right?

If you want to avoid moving the keys, shouldn't you exit early in `PSASessionKeystore::PersistICDKey` if the key is already persistent?

Reply:
> Hmm. Even if you derived the key ID from `index`, the shifting mechanism would work correctly because during the shift `ICDMonitoringTable::Set` would be called, which would _move_ the corresponding PSA keys into the new IDs, right?
> 
> If you want to avoid moving the keys, shouldn't you exit early in `PSASessionKeystore::PersistICDKey` if the key is already persistent?

Oh. I forgot to add exiting early for already present persistent keys.. Thanks for spotting that! The idea was to minimize moving of the keys, so with that fixed, only key ID will be copied, and the only way to change key itself should be through `SetKey` . With index derivation only we would need to add yet another crypto API for key copying.

Reply:
> With index derivation only we would need to add yet another crypto API for key copying.

Not if `PersistICDKey` would take the index and always moved the key to the new ID, but I'm OK with the current solution if the early exiting is added :).

Reply:
Added early exit in `PersistICDKey`